cmd/snap-bootstrap/hybrid_users_import.go

```diff
@@ -0,0 +1,462 @@
+	})
+	if err != nil {
+		return err
+	}
+
+	var groupBuffer bytes.Buffer
+	var gshadowBuffer bytes.Buffer
+	for name, group := range sourceGroups {
+		if name != group.name {
+			return errors.New("internal error: group entry inconsistent with parsed data")
+		}
+
+		// if we're importing a group that already exists, we take the imported
+		// group.
+		if _, ok := groups[name]; ok {
+			continue
+		}
+
+		// we combine the users that are already in the group with the new users
+		// that we're importing
+		usersInGroup := unique(append(group.users, groupsToNewUsers[name]...))
+		sort.Strings(usersInGroup)
+
+		usersPart := strings.Join(usersInGroup, ",")
+		parts := strings.Split(group.groupEntry, ":")
+
+		fmt.Fprintf(&groupBuffer, "%s:%s:%s:%s\n", parts[0], parts[1], parts[2], usersPart)
+
+		// if this group has a gshadow entry, we need to update the list of
+		// users there as well. not having a gshadow entry is fine, though.
+		if group.gshadowEntry == "" {
+			continue
+		}
+
+		gshadowParts := strings.Split(group.gshadowEntry, ":")
+		if len(gshadowParts) != 4 {
+			continue
+		}
+
+		fmt.Fprintf(&gshadowBuffer, "%s:%s:%s:%s\n", gshadowParts[0], gshadowParts[1], gshadowParts[2], usersPart)
+	}
+
+	// add the group entries for the users that we're importing. note that we
+	// don't need to mess with the users in these groups, since they're coming
+	// directly from the system we're importing from
+	for _, g := range groups {
+		if g.gid != 0 && g.gid < 1000 {
+			return fmt.Errorf("internal error: cannot import system groups: %s has gid %d", g.name, g.gid)
+		}
+
+		groupBuffer.WriteString(g.groupEntry + "\n")
+		if g.gshadowEntry != "" {
+			gshadowBuffer.WriteString(g.gshadowEntry + "\n")
+		}
+	}
+
+	destinationGroup := filepath.Join(outputDir, "group")
+	if err := os.WriteFile(destinationGroup, groupBuffer.Bytes(), 0644); err != nil {
+		return err
+	}
+
+	destinationGShadow := filepath.Join(outputDir, "gshadow")
+	return os.WriteFile(destinationGShadow, gshadowBuffer.Bytes(), 0600)
+}
+
+// unique returns a slice with unique entries; the provided slice is modified
+// and should not be re-used.
+func unique(slice []string) []string {
+	seen := make(map[string]bool)
+	current := 0
+	for _, entry := range slice {
+		if _, ok := seen[entry]; ok {
+			continue
+		}
+
+		seen[entry] = true
+		slice[current] = entry
+		current++
+	}
+	return slice[:current]
+}
+
+func parseUsers(root string, filter func(user) bool) (map[string]user, error) {
+	passwdEntries, err := entriesByName(filepath.Join(root, "etc/passwd"))
+	if err != nil {
+		return nil, err
+	}
+
+	shadowEntries, err := entriesByName(filepath.Join(root, "etc/shadow"))
+	if err != nil {
+		return nil, err
+	}
+
+	userToGroups, err := parseGroupsForUser(filepath.Join(root, "etc/group"))
+	if err != nil {
+		return nil, err
+	}
+
+	users := make(map[string]user)
+	for name, entry := range passwdEntries {
+		parts := strings.Split(entry, ":")
+		if len(parts) != 7 {
```

Comment:
The interpreter is optional (see passwd(5))

Reply:
I believe the colon will always be present, though. When that man page says optional, I believe it is talking about an empty string. This is consistent with the "encrypted password field".

Reply:
I made a quick check by adding an entry like `aa:aa:1005:1005:aa:/home/aa` and I can do `su - aa`, so it looks like not having the last colon is fine. But good point on the other field being optional as well.

Reply:
Okay, in that case I've pushed a change to allow entries that look like this.